django-stubs/db/models/query_utils.pyi

```diff
@@ -1,34 +1,53 @@
+import sys
 from collections import namedtuple
-from typing import Any, Collection, Dict, Iterator, List, Mapping, Optional, Sequence, Set, Tuple, Type
+from typing import (
+    Any,
+    Collection,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal
+else:
+    from typing import Literal
 
 from django.db.backends.base.base import BaseDatabaseWrapper
 from django.db.models.base import Model
+from django.db.models.expressions import BaseExpression
 from django.db.models.fields import Field
 from django.db.models.fields.mixins import FieldCacheMixin
+from django.db.models.lookups import Lookup, Transform
 from django.db.models.sql.compiler import SQLCompiler
 from django.db.models.sql.query import Query
 from django.db.models.sql.where import WhereNode
 from django.utils import tree
 
-PathInfo = namedtuple("PathInfo", "from_opts to_opts target_fields join_field m2m direct filtered_relation")
+PathInfo = namedtuple(
+    "PathInfo", ["from_opts", "to_opts", "target_fields", "join_field", "m2m", "direct", "filtered_relation"]
+)
 
 class InvalidQuery(Exception): ...
 
 def subclasses(cls: Type[RegisterLookupMixin]) -> Iterator[Type[RegisterLookupMixin]]: ...
 
-class QueryWrapper:
-    contains_aggregate: bool = ...
-    data: Tuple[str, List[Any]] = ...
-    def __init__(self, sql: str, params: List[Any]) -> None: ...
-    def as_sql(self, compiler: Optional[SQLCompiler] = ..., connection: Optional[BaseDatabaseWrapper] = ...) -> Any: ...
-
 class Q(tree.Node):
     AND: str = ...
     OR: str = ...
     conditional: bool = ...
-    def __init__(self, *args: Any, **kwargs: Any) -> None: ...
-    def __or__(self, other: Any) -> Q: ...
-    def __and__(self, other: Any) -> Q: ...
+    def __init__(self, *args: Any, _connector: Optional[Any] = ..., _negated: bool = ..., **kwargs: Any) -> None: ...
```

Comment:
Arguably this is correct, but it confuses mypy when passing dynamically generated dict as an argument: 

```
filter = Q(**{f"{self.field_name}__{self.lookup_expr}": value})
```

Causes mypy to say:

> error: Argument 1 to "Q" has incompatible type "**Dict[str, List[str]]"; expected "bool"

Given that these underscore arguments are used in Django internally and not public usage, maybe revert this change?

Reply:
Thanks! I agree that it's much more important.

Reply:
I ran into the issue as well. I wasn't sure my kind of usage was representative.